criu/tls.c

```diff
@@ -0,0 +1,336 @@
+#ifdef CONFIG_HAS_GNUTLS
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+#include <gnutls/gnutls.h>
+
+#include "cr_options.h"
+#include "xmalloc.h"
+
+#define TLS_PRIORITY 					\
+	"NORMAL" 					\
+	":+PSK:+ECDHE-PSK:+DHE-PSK"			\
+	":-VERS-SSL3.0:-VERS-TLS1.0:-VERS-TLS1.1"	\
+	":-MD5:-SHA1"
+
+#define tls_perror(msg, ret)	pr_err("%s: %s\n", msg, gnutls_strerror(ret))
+
+static gnutls_session_t session;
+static gnutls_priority_t priority_cache;
+static gnutls_psk_client_credentials_t c_psk_cred;
+static gnutls_psk_server_credentials_t s_psk_cred;
+
+static int tls_sk = -1;
+static int *connfdPtr = NULL;
+
+void tls_terminate_session()
+{
+	int ret;
+	do {
+		/* do not wait for the peer to close the connection */
+		ret = gnutls_bye(session, GNUTLS_SHUT_WR);
+	}
+	while(ret == GNUTLS_E_AGAIN || ret == GNUTLS_E_INTERRUPTED);
+	gnutls_deinit(session);
+}
+
+void tls_deinit()
+{
+	tls_sk = -1;
+	if (connfdPtr)
+		xfree(connfdPtr);
+	if (c_psk_cred)
+		gnutls_psk_free_client_credentials(c_psk_cred);
+	if (s_psk_cred)
+		gnutls_psk_free_server_credentials(s_psk_cred);
+	if (priority_cache)
+		gnutls_priority_deinit(priority_cache);
+	gnutls_global_deinit();
+}
+
+ssize_t tls_send(void *buf, size_t len)
+{
+	int ret;
+
+	do {
+		ret = gnutls_record_send(session, buf, len);
+	} while (ret == GNUTLS_E_INTERRUPTED || ret == GNUTLS_E_AGAIN);
+
+	if (gnutls_error_is_fatal(ret)) {
+		tls_perror("Faild to send data", ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+/*
+ * Read data from a file descriptor, then encrypt and send it with GnuTLS.
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from PIPE to TCP socket.
+ */
+ssize_t tls_send_data_from_fd(int fd, unsigned long len)
+{
+	ssize_t copied;
+	ssize_t total_sent = 0;
+	char *buffer = xmalloc(len);
```

Comment:
I wonder if we need to limit the size to be xmalloc()ed here.
Maybe something like (more question rather than suggestion):
```
#define SPLICE_BUF_SZ_MAX (PIPE_BUF * 100)  /* some reasonable limit */
[..]
char *buffer = alloca(min(len, SPLICE_BUF_SZ_MAX));
```